File: internal/tui/views/maintenance.go
package views

import (
	"fmt"
	"strings"

	"github.com/charmbracelet/lipgloss"

	"ash/internal/maintenance"
	"ash/internal/tui"
)

// MaintenanceView renders the maintenance commands screen.
type MaintenanceView struct {
	styles   tui.Styles
	commands []*maintenance.Command
	results  map[string]*maintenance.CommandResult
	cursor   int
	running  bool
	width    int
	height   int
}

// NewMaintenanceView creates a new maintenance view.
func NewMaintenanceView(styles tui.Styles) *MaintenanceView {
	return &MaintenanceView{
		styles:   styles,
		commands: maintenance.GetCommands(),
		results:  make(map[string]*maintenance.CommandResult),
	}
}

// SetSize sets the view dimensions.
func (v *MaintenanceView) SetSize(width, height int) {
	v.width = width
	v.height = height
}

// MoveUp moves the cursor up.
func (v *MaintenanceView) MoveUp() {
	if v.cursor > 0 {
		v.cursor--
	}
}

// MoveDown moves the cursor down.
func (v *MaintenanceView) MoveDown() {
	if v.cursor < len(v.commands)-1 {
		v.cursor++
	}
}

// SelectedCommand returns the currently selected command.
func (v *MaintenanceView) SelectedCommand() *maintenance.Command {
	if v.cursor >= 0 && v.cursor < len(v.commands) {
		return v.commands[v.cursor]
	}
	return nil
}

// Cursor returns the current cursor position.
func (v *MaintenanceView) Cursor() int {
	return v.cursor
}

// SetRunning sets whether a command is currently running.
func (v *MaintenanceView) SetRunning(running bool) {
	v.running = running
}

// SetResult sets the result for a command.
func (v *MaintenanceView) SetResult(cmd *maintenance.Command, result *maintenance.CommandResult) {
	v.results[cmd.Name] = result
}

// IsRunning returns whether a command is running.
func (v *MaintenanceView) IsRunning() bool {
	return v.running
}

// Render renders the maintenance view.
func (v *MaintenanceView) Render() string {
	var b strings.Builder

	b.WriteString("\n")
	b.WriteString(v.styles.Title.Render("Maintenance"))
	b.WriteString("\n")
	b.WriteString(v.styles.Subtitle.Render("System maintenance commands"))
	b.WriteString("\n\n")

	for i, cmd := range v.commands {
		line := v.renderCommand(cmd, i == v.cursor)
		b.WriteString(line)
		b.WriteString("\n")

		// Show result if available
		if result, ok := v.results[cmd.Name]; ok {
			b.WriteString(v.renderResult(result))
			b.WriteString("\n")
		}
	}

	b.WriteString("\n")
	b.WriteString(v.renderHelp())

	return b.String()
}

func (v *MaintenanceView) renderCommand(cmd *maintenance.Command, isCursor bool) string {
	cursor := "  "
	if isCursor {
		cursor = "> "
	}

	titleStyle := v.styles.FileName
	descStyle := v.styles.Subtitle

	if isCursor {
		bg := lipgloss.Color("#262626")
		titleStyle = titleStyle.Copy().Background(bg).Bold(true)
		descStyle = descStyle.Copy().Background(bg)
	}

	// Sudo indicator
	sudoIndicator := ""
	if cmd.RequiresSudo {
		sudoStyle := lipgloss.NewStyle().
			Foreground(lipgloss.Color("#f59e0b")).
			Bold(true)
		sudoIndicator = sudoStyle.Render(" [sudo]")
	}

	// Useful indicator
	usefulIndicator := ""
	if !cmd.Useful {
		usefulStyle := lipgloss.NewStyle().
			Foreground(lipgloss.Color("#525252"))
		usefulIndicator = usefulStyle.Render(" (limited benefit)")
	}

	var b strings.Builder
	b.WriteString(cursor)
	b.WriteString(titleStyle.Render(cmd.Name))
	b.WriteString(sudoIndicator)
	b.WriteString(usefulIndicator)
	b.WriteString("\n")
	b.WriteString("    ")
	b.WriteString(descStyle.Render(cmd.Description))

	return b.String()
}

func (v *MaintenanceView) renderResult(result *maintenance.CommandResult) string {
	var status string
	var style lipgloss.Style

	if result.Success {
		style = lipgloss.NewStyle().Foreground(lipgloss.Color("#22c55e"))
		status = fmt.Sprintf("    + Completed in %s", result.Duration.Round(1))
	} else {
		style = lipgloss.NewStyle().Foreground(lipgloss.Color("#ef4444"))
		errMsg := "unknown error"
		if result.Error != nil {
			errMsg = result.Error.Error()
		}
		status = fmt.Sprintf("    x Failed: %s", errMsg)
	}

	return style.Render(status)
}

func (v *MaintenanceView) renderHelp() string {
	help := "j/k:navigate | enter:run | esc:back | q:quit"
	return v.styles.StatusBar.Render(help)
}

File: ash-zig/src/views/maintenance.zig
const std = @import("std");
const ansi = @import("../tui/ansi.zig");
const theme = @import("../tui/theme.zig");
const keybinds = @import("../tui/components/keybinds.zig");
const maintenance = @import("../maintenance.zig");

pub const MaintenanceView = struct {
    cursor: usize = 0,
    results: [maintenance.commands.len]?maintenance.CommandResult = .{null} ** maintenance.commands.len,
    running: ?usize = null,
    style: theme.Style = theme.Style.init(theme.default_theme),

    pub fn init() MaintenanceView {
        return .{};
    }

    pub fn moveUp(self: *MaintenanceView) void {
        if (self.cursor > 0) {
            self.cursor -= 1;
        }
    }

    pub fn moveDown(self: *MaintenanceView) void {
        if (self.cursor < maintenance.commands.len - 1) {
            self.cursor += 1;
        }
    }

    pub fn selectedCommand(self: MaintenanceView) *const maintenance.Command {
        return &maintenance.commands[self.cursor];
    }

    pub fn setRunning(self: *MaintenanceView, index: ?usize) void {
        self.running = index;
    }

    pub fn setResult(self: *MaintenanceView, index: usize, result: maintenance.CommandResult) void {
        self.results[index] = result;
    }

    pub fn render(self: MaintenanceView, writer: anytype, width: u16, height: u16) !void {
        _ = height;

        try writer.writeAll("\n");
        try self.style.bold(writer);
        try self.style.primary(writer);
        try writer.writeAll("  System Maintenance\n\n");
        try self.style.reset(writer);

        // Command list
        for (maintenance.commands, 0..) |cmd, i| {
            const is_current = i == self.cursor;
            const is_running = self.running == i;
            const result = self.results[i];

            // Cursor
            if (is_current) {
                try self.style.primary(writer);
                try writer.writeAll("  > ");
            } else {
                try writer.writeAll("    ");
            }

            // Status indicator
            if (is_running) {
                try self.style.warning(writer);
                try writer.writeAll("[...] ");
            } else if (result) |r| {
                if (r.success) {
                    try self.style.success(writer);
                    try writer.writeAll("[OK]  ");
                } else {
                    try self.style.danger(writer);
                    try writer.writeAll("[ERR] ");
                }
            } else {
                try self.style.muted(writer);
                try writer.writeAll("      ");
            }

            // Command name
            if (is_current) {
                try self.style.bold(writer);
                try self.style.primary(writer);
            } else {
                try self.style.secondary(writer);
            }
            try writer.writeAll(cmd.name);

            // Sudo indicator
            if (cmd.requires_sudo) {
                try self.style.muted(writer);
                try writer.writeAll(" [sudo]");
            }
            try self.style.reset(writer);
            try writer.writeAll("\n");

            // Description
            try writer.writeAll("          ");
            try self.style.muted(writer);
            try writer.writeAll(cmd.description);
            try self.style.reset(writer);

            // Result details
            if (result) |r| {
                if (r.success) {
                    try self.style.success(writer);
                    const duration_ms = r.duration_ns / 1_000_000;
                    try writer.print(" (completed in {d}ms)", .{duration_ms});
                } else {
                    try self.style.danger(writer);
                    if (r.errorMsg()) |err| {
                        try writer.print(" - {s}", .{err});
                    }
                }
                try self.style.reset(writer);
            }

            try writer.writeAll("\n\n");
        }

        // Footer
        const bindings = [_]keybinds.KeyBind{
            .{ .key = "j/k", .description = "navigate" },
            .{ .key = "enter", .description = "run" },
            .{ .key = "esc", .description = "back" },
            .{ .key = "q", .description = "quit" },
        };
        try keybinds.renderFooter(writer, &bindings, width, self.style);
    }
};

// Tests
test "MaintenanceView" {
    var view = MaintenanceView.init();

    try std.testing.expectEqual(@as(usize, 0), view.cursor);

    view.moveDown();
    try std.testing.expectEqual(@as(usize, 1), view.cursor);

    view.moveUp();
    try std.testing.expectEqual(@as(usize, 0), view.cursor);
}
